Lib/functools.py

```diff
@@ -192,6 +192,208 @@ def total_ordering(cls):
+                self.add(node, *predecessors)
+
+    def _get_nodeinfo(self, node):
+        if (result := self.node2info.get(node)) is None:
+            self.node2info[node] = result = _NodeInfo(node)
+        return result
+
+    def add(self, node, *predecessors):
+        """Add a new node and its predecessors to the graph.
+
+        Both the *node* and all elements in *predecessors* must be hashable.
+
+        Raises ValueError if called after "prepare".
+        """
+        if self.ready_nodes is not None:
+            raise ValueError("Nodes cannot be added after a call to prepare()")
+
+        # Create the node -> predecessor edges
+        nodeinfo = self._get_nodeinfo(node)
+        nodeinfo.npredecessors += len(predecessors)
+
+        # Create the predecessor -> node edges
+        for pred in predecessors:
+            pred_info = self._get_nodeinfo(pred)
+            pred_info.successors.append(node)
+
+    def prepare(self):
+        """Mark the graph as finished and check for cycles in the graph.
+
+        If any cycle is detected, "CycleError" will be raised, but "get_ready" can
+        still be used to obtain as many nodes as possible until cycles block more
+        progress. After a call to this function, the graph cannot be modified and
+        therefore no more nodes can be added using "add".
+        """
+        if self.ready_nodes is not None:
+            raise ValueError("cannot prepare() more than once")
+
+        self.ready_nodes = [i.node for i in self.node2info.values()
+                            if i.npredecessors == 0]
+        # ready_nodes is set before we look for cycles on purpose:
+        # if the user wants to catch the CycleError, that's fine,
+        # they can continue using the instance to grab as many
+        # nodes as possible before cycles block more progress
+        cycle = self._find_cycle()
+        if cycle:
+            raise CycleError(f"nodes are in a cycle", cycle)
+
+    def get_ready(self):
+        """Return a tuple of all the nodes that are ready.
+
+        Initially it returns all nodes with no predecessors and once those are marked
+        as processed by calling "done", further calls will return all new nodes that
+        have all their predecessors already processed until no more progress can be
+        made.
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        # Get the nodes that are ready and mark them
+        result = tuple(self.ready_nodes)
+        n2i = self.node2info
+        for node in result:
+            n2i[node].npredecessors = _NODE_OUT
+
+        # Clean the list of nodes that are ready and update
+        # the counter of nodes that we have returned.
+        self.ready_nodes.clear()
+        self.npassedout += len(result)
+
+        return result
+
+    def is_active(self):
```

Comment:
Sorry if I'm late to the party.  But--why is this not named __bool__?

Reply:
I thought about it and had it in my first version but when I showed that to some people they were a big confused (as there are some rules on when this returns true) so almost all them prefer a explicit method.

We additionally add __bool__ that calls this.

Any opinions on this @tim-one?

Reply:
@larryhastings Added `__bool__` in 6f33c71

Reply:
Since the input to a topsort is a graph, my initial _guess_ would be that `__bool__` told me whether or not the graph was still in its initial empty state.  Which is pretty darned useless :wink:.

I don't particularly object to adding it to mean `is_active()`, but I'll never use it for that - too magical for something I'll use only rarely.  Note that in the doc examples it's used only once each in the serial and parallel cases.  I have fancier ways to use it (e.g., for parallel use with priority queues that delay passing out anything until a worker requests a new piece of work to do), and those ways `is_active()` can be called from multiple places - but then those are so much more involved that I find it a positively good thing that `is_active()` is explicitly spelled out.

In any case, if `__bool__()` is added to the code then the docs and tests have to change too to reflect that.

Reply:
I felt it provided a natural UX:

`while graph:
    for node in graph.ready():
        node.do_stuff()
        graph.done(node)`

That's how I interact with my eerily-similar `DependencyGraph()` data structure.

But I don't have a strong feeling about it, and if people hate using `__bool__` here then I won't fight for it.  I guess I'd put myself down as a +0.5, which on my project of one person is entirely sufficient to get the feature in ;-)

Reply:
Let me add that in my implementation, `__bool__` doesn't care whether or not the graph has been prepared (or "finalized" in my version).  I saw it as being like bool(list) or bool(dict): does the graph contain anything?  Though in this case it specifically means does it contain *work.*  So bool(graph) is True if it contains unprocessed nodes, and bool(graph) is False if all the nodes have been processed, or the graph is empty.  This is sufficiently different from conventional collections to give one pause--though it wasn't a concern for me working on my small project.

Reply:
@larryhastings @tim-one, By the way, what do you think about adding `__next__` so you can do:

```
for node_group in ts:
   # Process in serial/parallel nodes in node_group
    ts.done(*node_group)
```

basically would be a dead-simple wrapper over the `is_active` and `get_ready` methods but simplify the normal case. What do you think? Do you see any disadvantages? (Think of it as a way to simplify the basic case.

Reply:
The first thing I'll say is: `__next__` returning iterables of nodes is a little weird.  I understand why you did it that way.  But the UI you really _want_ is returning individual nodes:
```
graph.prepare()
for node in graph:
  process(node)
  graph.done(node)

```
And honestly I think that's the most desirable, most obvious UI.  But I didn't do that in my implementation because it doesn't work well with paralellism.

In your proposal above, you process nodes in batches.  The first time through the loop, it returns N nodes ready to process.  You process all N nodes, you wait for all N nodes to complete, then you mark all N nodes as done.  Now you loop and this time there are X nodes ready to process.  You do it all again, and you keep going until it raises `StopIteration`.  Good so far.

But what if you get two nodes in the first batch, A and B, and A takes 1 unit of time to process, but B takes 500 units of time?  And what if there are a bunch of nodes that are only dependent on A.  Now you're sitting around waiting for the first node to complete when you could be running more jobs.

When processing nodes in parallel, it only seems sensible to me to mark nodes as "done" _as soon as their job completes._  Then, every time you mark a node as done, you check to see if there are new ready nodes, and if so you can start processing them immediately--up to the limits of your job pool or whatnot.

So what you really want is more like this:
```
graph.prepare()
for node in graph:
    process_in_parallel(node)
```
And the parallel process, whatever it is, calls `graph.done(node)`.

You might be thinking "now you have a locking problem".  Well, that's true, but that's solveable.  But there's another, much harder problem to solve: what should `__next__` do when there are no "ready" nodes?  Blocking is terrible, but raising `StopIteration` is even more terrible.

The whole point of this slightly-strange API design is that it makes its agnostic about serial vs. parallel processing.  The cost of that is that we don't get the slick, obvious "iterate over the graph" API that everybody would normally expect.

Reply:
In thinking about it more, we could solve the problem if

1. we stick with returning an iterable of nodes as you suggest, and
2. if there are no ready nodes, we return an empty iterable.

This makes the API pretty strange.  And it's really no more convenient to use than the current API.  We still have nested for loops.

Current:
```
graph.prepare()
while graph:
  for node in graph.ready():
    process(node)
    graph.done(node)
```

Proposed:
```
graph.prepare()
for node_iterable in graph:
  for node in node_iterable:
    process(node)
    graph.done(node)
```

Personally I don't find the second particularly preferable over the first.

@tim-one you have an opinion here?

Reply:
> 2\. f there are no ready nodes, we return an empty iterable.

In which case, the following `for node in node_iterable` executes in nanoseconds and so we almost immediately poke the iterator again.  But no progress has been made (no new nodes have been marked done), so it will return an empty iterable again.  Lather,  rinse, repeat.  That is, if case 2 ever obtains, it's an infinite loop.

All ways of making an instance an iterable suck in the face of parallelism.

But not all users will care about that, and there's nothing wrong with adding a method that returns a one-and-done iterable, not requiring (or likely even allowing) calls to prepare(), get_ready(), done(), or is_active():

```python
def static_order(self):
    self.prepare()
    while self.is_active():
        chunk = self.get_ready()
        yield from chunk
        self.done(*chunk)
```

I do resist magic beyond that.  Users aren't typically "topsort experts", or likely to use this so often that magical shortcuts help more than they obscure.  If all they want is _a_ topsort, fine, they only need to learn about add() and one method to generate a topsort order.

The docs could usefully show the implementation of that "easy peasy" method, to drive home what the lower-level methods do.

Reply:
I am convinced! I think for most users adding the `static_order` method can be useful with the documentation additions that @tim-one mentions. I have added such a method in aa56145a04.